pharus/component_interface.py

```diff
@@ -114,8 +109,167 @@ def restriction(self):
+        self.type = component_config["type"]
+        self.route = component_config["route"]
+        self.connection = dj.conn(
+            host=jwt_payload["databaseAddress"],
+            user=jwt_payload["username"],
+            password=jwt_payload["password"],
+            reset=True,
+        )
+        self.fields_map = component_config["map"] if "map" in component_config else None
+        self.tables = [
+            getattr(
+                dj.VirtualModule(
+                    s,
+                    s,
+                    connection=self.connection,
+                ),
+                t,
+            )
+            for s, t in (_.split(".") for _ in component_config["tables"])
+        ]
+
+    def _get_datatype_map(self):
+        attributes = reduce(
+            lambda a0, a1: {**a0, **a1}, (t.heading.attributes for t in self.tables)
+        )
+        datatype_map = {k: v.type for k, v in attributes.items()}
+        return datatype_map
 
-class TableComponent(QueryComponent):
+    def dj_query_route(self):
+        with self.connection.transaction:
+            if self.fields_map:
+                field_name_map = {}
+                for field in self.fields_map:
+                    if field["type"] != "table":
+                        field_name_map[
+                            field["input"] if "input" in field else field["destination"]
+                        ] = field["destination"]
+                    else:
+                        for submap in field["map"]:
+                            field_name_map[
+                                submap["input"]
+                                if "input" in submap
+                                else submap["destination"]
+                            ] = submap["destination"]
+                self.payload = {field_name_map[k]: v for k, v in self.payload.items()}
+            for t in self.tables:
+                t.insert1(
+                    {k: v for k, v in self.payload.items() if k in t.heading.attributes}
+                )
+        return "Insert successful"
+
+    def fields_route(self):
+        fields = []
+        if self.fields_map:
+            datatype_map = self._get_datatype_map()
+            for field in self.fields_map:
+                field_type = field["type"]
+                field_name = (
+                    field["input"] if "input" in field else field["destination"]
+                )
+                if field_type == "attribute":
+                    field_datatype = datatype_map[field["destination"]]
+                    fields.append(
+                        dict(type=field_type, datatype=field_datatype, name=field_name)
+                    )
+                elif field_type == "table":
+                    s, t = field["destination"].split(".")
+                    if "map" in field:
+                        proj_map = {
+                            attr["input"]
+                            if "input" in attr
```

Comment:
This deep nested if's are a bit hard to follow the logic. A minor point but still took me a bit to understand it on first glance.